jsonrpc/event_listener.go

```diff
@@ -1,9 +1,14 @@
 package jsonrpc
 
-import "time"
+import (
+	"net"
+	"time"
+)
 
 type NewRequestListener interface {
 	OnNewRequest(method string)
+	OnNewConnection(conn net.Conn)
```

Comment:
It seems dangerous to allow `NewRequestListener` implementations to arbitrarily read, write, or close the connection. Do you think it is sufficient to pass in `remoteAddr string` and then include that as a label in our metrics?

Reply:
I don't think prometheus will be happy about having so many unique labels and it will be super hard for visualization in grafana. I think that `remoteAddr` should fall into logging instead and then it should be implemented for every handler instead.

Reply:
If it's too dangerous then I think that it can be further split into:
- OnNewConnection(), OnDisconnect() - metric for amount of connections. could even be done for all handlers (http,ws,...)
- OnRead, OnWrite - once again, for all handlers - super usefull for tracking egress/ingress since it's very expensive of `aws/gc`

Reply:
> OnNewConnection(), OnDisconnect() - metric for amount of connections. could even be done for all handlers (http,ws,...)

I think `OnNewConnection()`/`OnDisconnect` makes sense for websockets and ipc. Are they redundant with `OnNewRequest`,etc. for http?

> OnRead(amount uint), OnWrite(amount uint)

This is a great idea. We need to work some details out. For example, we read from the connection using `dec.Decode`, which doesn't tell us how much it reads; it may even read past the json request object. Do you mind creating a separate issue for this?

Reply:
I take the point about `decoder.Decode` back: it looks like we can use its `InputOffset` function to determine how much it has read. Regardless, let's reserve the `OnRead`/`OnWrite` idea for a separate PR.

Reply:
> I think OnNewConnection()/OnDisconnect makes sense for websockets and ipc. Are they redundant with OnNewRequest,etc. for http?

Not quite since http connection can reuse the same socket. Simply counting `StateNew` 
 and `StateClosed` from that returns the actual amount which can be lower than `OnNewRequest` but never higher
 https://github.com/golang/go/blob/d95e25e83c922aa63fcb1f596d6bdf1786789edb/src/net/http/server.go#L2890C2-L2890C37 
If it's worth it? not sure